gpt_json/gpt.py

```diff
@@ -144,6 +139,58 @@ async def run(
         else:
             return self.schema_model(**extracted_json), fixed_payload
 
+    async def stream(
+        self,
+        messages: list[GPTMessage],
+        max_response_tokens: int | None = None,
+        format_variables: dict[str, Any] | None = None,
+    ) -> AsyncGenerator[StreamingObject[SchemaType], None]:
+        """
+        See `run` for documentation. This method is an async generator wrapper around `run` that streams partial results
+        instead of returning them all at once.
+
+        :return: yields `StreamingObject[SchemaType]`s. 
+        """
+        if self.extract_type != ResponseType.DICTIONARY:
+            raise NotImplementedError("For now, streaming is only supported for dictionary responses.")
+        for field_type in self.schema_model.__annotations__.values():
+            if field_type != str:
+                raise NotImplementedError("For now, streaming is only supported for dictionary responses with string fields.")
```

Comment:
I think list values are reasonable to scope out to another PR.

But before limiting this PR to only string values, I'd like to sketch out what the expected behavior for other value types might look like. Some questions:

- If a number is being sent, presumably the streamer should wait until the whole number is collected before sending it back to the caller?
- If a list is being sent, should it yield after every individual value is completed? If we know that it should be yielding a list of strings, should we send those incrementally? (Probably yes) But lists of numbers should probably be queued at least until each value is completed. So there's some compositionally to the expected yielding based on the type value that is currently being processed.

Concretely, let's write up a few simple test cases with different schemas and different expected streaming patterns? We can go to the implementation from there depending on complexity. Let's discuss further in the original issue.

Reply:
hm maybe for a list, `updated_key = 0, 1, 2, etc`. for arbitrarily nested objects, let's a list of schema objects, `updated_key = (0, "schema_key_1"), (0, "schema_key_2"), (1, "schema_key_1"), etc"`

Reply:
took a stab at testcases for singly-nested response streaming with arbitrary leaf types (see `gpt_json/tests/test_streaming.json`) -- I still need to test `gpt_json.stream` by mocking OAI completionchunks etc, but I wanted to check in to see what you think of the streaming testcases first